Adding support for nil values

alexandergriekspoor opened the pull request.

alexandergriekspoor committed:
Added support for nil values in PARStore

alexandergriekspoor committed:
Updated xcode project

drewmccormack commented on Core/PARStore.m:
```diff
@@ -768,7 +774,7 @@ - (NSData *)dataFromPropertyList:(id)plist error:(NSError **)error
 - (id)propertyListFromData:(NSData *)blob error:(NSError **)error
 {
-    if (!blob)
+    if (!blob || blob.length == 0)
```
I guess one thing we may have lost here is the ability to store an empty data, right? Will now always become nil. Does raise the question of whether some specific data value for null might not be better. Eg. some string converted to data.

alexandergriekspoor replied:
We decided it was fine to use an empty data as placeholder as a real empty NSData property list value would be still encoded as a plist and therefore correctly discriminated from the placeholder

cparnot replied:
Addressed

drewmccormack commented on Core/PARStore.m:
```diff
@@ -921,7 +925,11 @@ - (void)setEntriesFromDictionary:(NSDictionary *)dictionary timestampApplied:(NS
     [self.memoryQueue dispatchSynchronously:^
      {
-         [self._memory addEntriesFromDictionary:dictionary];
+         // each key/value --> add to memory story if the value is not a marker for a removed value
```
Small typo on 'story'

alexandergriekspoor replied:
Fixed

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1398,9 +1407,15 @@ - (void)applySyncChangeWithValues:(NSDictionary *)values timestamps:(NSDictionar
-        [values enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *s)     { self._memory[key] = obj;              }];
+        [values enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *s)
+        {
+            self._memory[key] = (obj != [NSNull null] ? obj : nil);
```
So if the obj is Null, we remove the key from the memory dictionary? Isn't that a problem for whatever is using the PARStore? Doesn't it need to know the difference between something that has been set to nil explicitly, and something which has never been set? Don't we have to carry that NSNull through the whole system in some form? In the case of Agenda, my guess is that NSNull would become an AnyObject? with a nil value.

alexandergriekspoor replied:
Conclusion from our Skype call is that this is the desired behaviour

cparnot replied:
Addressed

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1579,7 +1595,10 @@ - (void)_sync
-                 self._memory = [NSMutableDictionary dictionaryWithDictionary:updatedValues];
+                 [updatedValues enumerateKeysAndObjectsUsingBlock:^(id key, id newValue, BOOL *stop)
+                 {
+                     self._memory[key] = (newValue != [NSNull null] ? newValue : nil);
```
Same question: Is removing the key the right answer here? We are discarding information that could be important to the caller.

alexandergriekspoor replied:
See above

drewmccormack reviewed (changes requested):
I have a couple of concerns:

1. Using an empty data as nil means we can no longer represent a true empty data valued property, which is probably not that rare. I would prefer we pick an obscure Data value to represent nil. One that won't collide with property values.
2. It seems like NSNull values cause in memory data to be removed altogether. I would think the NSNull should remain in memory, so the calling code can distinguish between something never added, and something removed. Ie we need to propagate the explicit nil state through the whole system.

cparnot committed:
Adding a test for setting nil to a property.

For now, the test fails. In fact, the code crashes because you can’t
remove a value using the ObjC subscript syntax with a nil value.

alexandergriekspoor committed:
Fixed issue with objc subscripting not supporting nil

alexandergriekspoor committed:
Fixed a nil value dictionary exception

alexandergriekspoor commented:
I fixed the initial issue with the nil setter test after fixing the subscripting's inability to take nil. Unfortunately it revealed another issue, apparently giving a zero-length data object to CoreData makes it consider like no data was given at all (i.e. nil). We'll need to look into whether this is a problem (i.e. can we work with nil or do we have to switch to a different placeholder

cparnot committed:
Fully fixing the handling of nil in the individual key-value setter implementation.

There were various issues that would lead to crashes or exception with
the behavior of ‘nil’. Sorry, Alex, I am rewriting a bit of what you
fixed already, but that’s based on the test results.

cparnot committed:
Fixing the handling of nil in the load/sync implementation.

The main issue was that a nil blob would be considered an error, where
in fact saving [NSData data] into a CoreData column of type data is
apparently equivalent to setting the value to nil, and we should thus
instead consider a nil value as a removal marker.

cparnot committed:
Adding a test for using `NSNull` as the removal marker in the implementation of the dictionary setter.

Test is passing.

cparnot committed:
Simpler code for initialization of the memory cache on first load.

No need to set a value to nil when we know the dictionary does not have
any value for any of the keys yet.

cparnot committed:
Adding tests for the use of the method `insertChanges:forDeviceIdentifier:appendOnly:error:`.

One of the test also uses a nil value to test it’s correctly marked for
removal.

cparnot committed:
Adding assertion instead of a defensive `removeAllObjects` call that could mask bugs down the road.

cparnot committed:
Merge branch 'master' into ag-null

This applies a bug fix important to get all the tests to pass and make
sure they still pass with the ability to set a property list to nil.
There are a few more tests now, and they all pass, except the “merge”
tests that trigger an assertion.

cparnot committed:
Merge branch 'master' into ag-null

cparnot committed:
Enabling all tests again.

cparnot committed:
Adding a test for store merge that include a nil value.

cparnot commented:
I have merged the other recent changes from master, and added a number of tests for the use of nil values in the following context:

- simple setter
- simple setter + sync (with 2 devices)
- dictionary setter
- the `insertChanges:...`, used as an "external" setter when changes come from e.g. a separate iCloud sync setup
- merging stores

All the tests pass, and it looks like we have addressed all the issues and concerned raised during our discussion.

cparnot committed:
Merge branch 'master' into ag-null

# Conflicts:
#	Core/PARStore.m

cparnot committed:
Merge with master.

drewmccormack commented on Tests/PARStoreTests.m:
```diff
@@ -190,6 +190,32 @@ - (void)testPropertyListSetter
+    XCTAssertNil(document2.first, @"unexpected 'first' value: '%@' instead of nil", document2.first);
+    [document2 tearDownNow];
+    document2 = nil;
+}
```
Not really that important, but the `loadNow` and `tearDownNow` may be pointing to these tests being a bit too broad. Usually, a single standard `setup` and `teardown` should be enough for a single test. So instead of `testNilSetter`, you might have `testSavingAndLoadingDataWithNil` and `testLoadingDataWithNilFromSecondDocument`.

I wouldn't change anything now, but maybe worth thinking about for future tests.
